File: test/integ/CopyPropagationTest.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <gtest/gtest.h>

#include "ControlFlow.h"
#include "DexClass.h"
#include "DexInstruction.h"
#include "DexLoader.h"
#include "DexUtil.h"
#include "IRCode.h"
#include "RedexTest.h"
#include "Show.h"
#include "Transform.h"

#include "CopyPropagationPass.h"

int count_sgets(cfg::ControlFlowGraph& cfg) {
  int sgets = 0;
  for (auto& mie : InstructionIterable(cfg)) {
    TRACE(RME, 1, "%s", SHOW(mie.insn));
    if (opcode::is_an_sget(mie.insn->opcode())) {
      sgets++;
    }
  }
  return sgets;
}

class CopyPropagationTest : public RedexIntegrationTest {};

TEST_F(CopyPropagationTest, useSwitch) {
  TRACE(RME, 1, "Code before:");
  for (const auto& cls : *classes) {
    TRACE(RME, 1, "Class %s", SHOW(cls));
    for (const auto& m : cls->get_vmethods()) {
      TRACE(RME, 1, "\nmethod %s:", SHOW(m));
      IRCode* code = m->get_code();
      code->build_cfg();
      EXPECT_EQ(2, count_sgets(code->cfg()));
      code->clear_cfg();
    }
  }

  auto copy_prop = new CopyPropagationPass();
  copy_prop->m_config.static_finals = true;
  std::vector<Pass*> passes = {copy_prop};
  run_passes(passes);

  TRACE(RME, 1, "Code after:");
  for (const auto& cls : *classes) {
    for (const auto& m : cls->get_vmethods()) {
      TRACE(RME, 1, "\nmethod %s:", SHOW(m));
      IRCode* code = m->get_code();
      code->build_cfg();
      if (strcmp(m->get_name()->c_str(), "remove") == 0) {
        EXPECT_EQ(1, count_sgets(code->cfg()));
      } else {
        EXPECT_EQ(2, count_sgets(code->cfg()));
      }
      code->clear_cfg();
    }
  }
}

File: test/integ/CopyPropagationTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.redextest;

import java.util.Random;

class CopyPropagationTest {

  public static final Object a = new Object();
  void remove(int p) {
    Object i = a;
    i = a;
  }

  public static Object b = new Object();
  void no_remove() {
    Object i = b;
    i = b; // should not be deleted because b is not final
  }

  public static final long c = new Random().nextLong();
  void remove(long p) {
    long l = c;
    l = c;
  }

  public static final double d = new Random().nextDouble();
  void remove(double p) {
    double n = d;
    int i = 0;
    if (new Random().nextInt() == 1) {
      i = 1;
    } else {
      i = 2;
    }
    n = d;
  }

  public static volatile int e = new Random().nextInt();
  void no_remove(int p) {
    int i = e;
    i = e;
  }
}

class Foo {
}
